packages/jsii/lib/assembler.ts

```diff
@@ -397,23 +405,49 @@ export class Assembler implements Emitter {
       node = type.symbol.declarations[0];
     }
 
-    const groups = /^"([^"]+)"\.(.*)$/.exec(tsName);
-    if (!groups) {
+    // Set to true to prevent further adding of Error diagnostics for known-bad reference
+    let hasError = false;
+
+    if (this._isPrivateOrInternal(type.symbol)) {
+      // Check if this type is "this" (explicit or inferred method return type).
+      const commonMessage = `cannot be used as the ${typeUse} because it is private or @internal`;
       this._diagnostic(
-        node,
+        resolvingNode,
         ts.DiagnosticCategory.Error,
-        `Cannot use private type ${tsName} in exported declarations`,
+        isThisType
+          ? `Type "this" (aka: "${type.symbol.name}") ${commonMessage}`
+          : `Type "${type.symbol.name}" ${commonMessage}`,
+        makeCause(node),
```

Comment:
I'm not sure I understand what the difference is between `resolvingNode` and `node`; would you want to use `resolvingNode` here in the `makeCause` call?

Reply:
Let me try a different name that just occurred to me...